File: 0016_3SumClosest/3SumClosest.cxx
#include<vector>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int ans;
        int Min = INT_MAX;
        auto num = nums.begin();
        while(num != nums.end()){
            auto lo = num + 1;
            auto hi = nums.end()-1;
            while(lo < hi){
                int tmp;
                int sum = *num + *lo + *hi; 
                if(  sum > target){
                    tmp = sum - target;
                    --hi;
                } 
                else if(sum < target) {
                    tmp = target - sum;
                    ++lo; 
                }
                else return target;
                if( tmp < Min ) {
                    ans = sum;
                    Min = tmp;
                }
            }
            ++num;
        }
        return ans;
    }
};

File: 0016_3SumClosest/3SumClosest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        Min = float("inf") 
        ans = float("inf") 

        for i in range(len(nums)):
            lo = i+1;
            hi = len(nums)-1
            while lo < hi:
                tmp=0
                S = nums[i]+nums[lo]+nums[hi];
                if S>target:
                    tmp = S-target
                    hi-=1
                elif S<target:
                    tmp = target-S
                    lo+=1
                elif S == target:
                    return target
                if tmp < Min:
                    Min=tmp
                    ans=S
        return ans